lib/tar_gz.ml

```diff
@@ -30,83 +80,86 @@ module Make
+        let rec until_full_or_end gz (res, res_off, res_len) =
           match Gz.Inf.decode gz with
           | `Flush gz ->
-            let max = Cstruct.length oc_buffer - Gz.Inf.dst_rem gz in
-            let len = min (Cstruct.length res) max in
-            Cstruct.blit oc_buffer 0 res 0 len ;
+            let max = De.bigstring_length oc_buffer - Gz.Inf.dst_rem gz in
+            let len = min res_len max in
+            bigstring_blit_bytes oc_buffer ~src_off:0 res ~dst_off:res_off ~len; 
             if len < max
             then ( state.pos <- len
                  ; state.gz <- gz
                  ; Async.return () )
-            else until_full_or_end (Gz.Inf.flush gz) (Cstruct.shift res len)
+            else until_full_or_end (Gz.Inf.flush gz) (res, res_off + len, res_len - len)
           | `End gz ->
-            let max = Cstruct.length oc_buffer - Gz.Inf.dst_rem gz in
-            let len = min (Cstruct.length res) max in
-            Cstruct.blit oc_buffer 0 res 0 len ;
-            if Cstruct.length res > len
+            let max = De.bigstring_length oc_buffer - Gz.Inf.dst_rem gz in
+            let len = min res_len max in
+            bigstring_blit_bytes oc_buffer ~src_off:0 res ~dst_off:res_off ~len; 
+            if res_len > len
             then raise End_of_file
             else ( state.pos <- len
                  ; state.gz <- gz
                  ; Async.return () )
           | `Await gz ->
-            Reader.read in_channel ic_buffer >>= fun len ->
-            let { Cstruct.buffer; off; len= _; } = ic_buffer in
-            let gz = Gz.Inf.src gz buffer off len in
-            until_full_or_end gz res
+            Reader.read in_channel tp_buffer >>= fun len ->
+            bigstring_blit_string (Bytes.unsafe_to_string tp_buffer) ~src_off:0 ic_buffer ~dst_off:0 ~len;
+            let gz = Gz.Inf.src gz ic_buffer 0 len in
+            until_full_or_end gz (res, res_off, res_len)
           | `Malformed err -> failwith ("gzip: " ^ err) in
-        let max = (Cstruct.length oc_buffer - Gz.Inf.dst_rem state.gz) - state.pos in
-        let len = min (Cstruct.length res) max in
-        Cstruct.blit oc_buffer state.pos res 0 len ;
-
+        let max = (De.bigstring_length oc_buffer - Gz.Inf.dst_rem state.gz) - state.pos in
+        let len = min (Bytes.length res) max in
+        bigstring_blit_bytes oc_buffer ~src_off:state.pos res ~dst_off:0 ~len; 
         if len < max
         then ( state.pos <- state.pos + len
              ; Async.return () )
-        else ( let res = Cstruct.shift res len in
-               until_full_or_end (Gz.Inf.flush state.gz) res )
+        else until_full_or_end (Gz.Inf.flush state.gz) (res, len, Bytes.length res - len)
 
     let skip
       : in_channel -> int -> unit io
       = fun state len ->
-        let oc_buffer = Cstruct.create len in
-        really_read state oc_buffer
+        let res = Bytes.create len in
+        really_read state res
```

Comment:
```suggestion
        Reader.skip state len
```

Reply:
Unfortunately, we can not skip like that through `Tar_gz`. If we want to skip N byte(s) (to go to the next item), it's N _decompressed_ byte(s) that we want to skip (not N _real_ byte(s)). So we must deflate and call the `really_read` implemented above.